mmedit/models/backbones/sr_backbones/basicvsr_pp.py

```diff
@@ -0,0 +1,409 @@
+            Default: True.
+        spynet_pretrained (str): Pre-trained model path of SPyNet.
+            Default: None.
+        cpu_cache_length (int): When the length of sequence is larger than
+            this value, the intermediate features are sent to CPU. This saves
+            GPU memory, but slows down the inference speed. You can increase
+            this number if you have a GPU with large memory. Default: 100.
+    """
+
+    def __init__(self,
+                 mid_channels=64,
+                 num_blocks=7,
+                 low_res_input=True,
+                 spynet_pretrained=None,
+                 cpu_cache_length=100):
+
+        super().__init__()
+
+        self.mid_channels = mid_channels
+        self.low_res_input = low_res_input
+        self.cpu_cache_length = cpu_cache_length
+
+        # optical flow
+        self.spynet = SPyNet(pretrained=spynet_pretrained)
+
+        # feature extraction module
+        if low_res_input:
+            self.feat_extract = ResidualBlocksWithInputConv(3, mid_channels, 5)
+        else:
+            self.feat_extract = nn.Sequential(
+                nn.Conv2d(3, mid_channels, 3, 2, 1),
+                nn.LeakyReLU(negative_slope=0.1, inplace=True),
+                nn.Conv2d(mid_channels, mid_channels, 3, 2, 1),
+                nn.LeakyReLU(negative_slope=0.1, inplace=True),
+                ResidualBlocksWithInputConv(mid_channels, mid_channels, 5))
+
+        # propagation branches
+        self.deform_align = nn.ModuleDict()
+        self.backbone = nn.ModuleDict()
+        modules = ['backward_1', 'forward_1', 'backward_2', 'forward_2']
+        for i, module in enumerate(modules):
+            self.deform_align[module] = SecondOrderDeformableAlignment(
+                2 * mid_channels, mid_channels, 3, padding=1, deform_groups=16)
+            self.backbone[module] = ResidualBlocksWithInputConv(
+                (2 + i) * mid_channels, mid_channels, num_blocks)
+
+        # upsampling module
+        self.reconstruction = ResidualBlocksWithInputConv(
+            5 * mid_channels, mid_channels, 5)
+        self.upsample1 = PixelShufflePack(
+            mid_channels, mid_channels, 2, upsample_kernel=3)
+        self.upsample2 = PixelShufflePack(
+            mid_channels, 64, 2, upsample_kernel=3)
+        self.conv_hr = nn.Conv2d(64, 64, 3, 1, 1)
+        self.conv_last = nn.Conv2d(64, 3, 3, 1, 1)
+        self.img_upsample = nn.Upsample(
+            scale_factor=4, mode='bilinear', align_corners=False)
+
+        # activation function
+        self.lrelu = nn.LeakyReLU(negative_slope=0.1, inplace=True)
+
+    def check_if_mirror_extended(self, lqs):
+        """Check whether the input is a mirror-extended sequence.
+
+        If mirror-extended, the i-th (i=0, ..., t-1) frame is equal to the
+        (t-1-i)-th frame.
+
+        Args:
+            lqs (tensor): Input LR images with shape (n, t, c, h, w)
+        """
+
+        self.is_mirror_extended = False
```

Comment:
This behavior is not elegant. We should set the public attribute for a class in its initialization functions. Can we indicate this attribute in the initialization function in advance?

Reply:
We have thought about it before in BasicVSR. In this setting, the model will check whether the input sequence is augmented. I designed it like this because users do not need to rebuild the model, they need to change only the data.

Reply:
I will update the model regarding the other comments first, can continue discussing this issue here.

Reply:
It's ok to keep this line here. I merely mean that we should set this attribute in the initialization function first.